clar2wasm/src/words/equal.rs

```diff
@@ -88,6 +89,179 @@ impl Word for IsEq {
     }
 }
 
+#[derive(Debug)]
+pub enum IndexOf {
+    Original,
+    Alias,
+}
+
+impl Word for IndexOf {
+    fn name(&self) -> ClarityName {
+        match self {
+            IndexOf::Original => "index-of".into(),
+            IndexOf::Alias => "index-of?".into(),
+        }
+    }
+
+    fn traverse(
+        &self,
+        generator: &mut WasmGenerator,
+        builder: &mut walrus::InstrSeqBuilder,
+        _expr: &SymbolicExpression,
+        args: &[SymbolicExpression],
+    ) -> Result<(), GeneratorError> {
+        // Traverse the sequence, leaving its offset and size on the stack.
+        let seq = args.get_expr(0)?;
+        generator.traverse_expr(builder, seq)?;
+        // STACK: [offset, size]
+
+        // Drop the sequence size from the stack.
+        builder.drop();
+        // STACK: [offset]
+
+        // Move the sequence offset from the stack to a local variable.
+        let offset = generator.module.locals.add(ValType::I32);
+        builder.local_set(offset);
+        // STACK: []
+
+        // Get the type of the sequence.
+        let seq_ty = match generator
+            .get_expr_type(seq)
+            .expect("sequence must be typed")
+        {
+            TypeSignature::SequenceType(seq_ty) => seq_ty.clone(),
+            _ => {
+                return Err(GeneratorError::InternalError(
+                    "expected sequence type".to_string(),
+                ));
+            }
+        };
+
+        // Get sequence size (quantity of items in the sequence)
+        // and sequence elements type.
+        let (seq_size, elem_ty) = match &seq_ty {
+            SequenceSubtype::ListType(list_type) => {
+                (list_type.get_max_len(), list_type.get_list_item_type())
+            }
+            // TODO implement check for other sequence types
+            _ => unimplemented!("Unsupported sequence type"),
+        };
+
+        // TODO return when size = 0
+
+        // Store the end of the sequence into a local.
+        let elem_size = get_type_size(elem_ty);
+        let end_offset = generator.module.locals.add(ValType::I32);
+        builder
+            .local_get(offset)
+            .i32_const(seq_size as i32 * elem_size)
+            .binop(BinaryOp::I32Add)
+            .local_set(end_offset);
+
+        // Traverse the item, leaving it on top of the stack.
+        let item = args.get_expr(1)?;
+        generator.traverse_expr(builder, item)?;
+        // STACK: [item]
+
+        // Get the type of the item expression
+        let item_ty = generator
+            .get_expr_type(item)
+            .expect("index_of item expression must be typed")
+            .clone();
+
+        // Store the item into a local.
+        let item_locals = generator.save_to_locals(builder, &item_ty, true);
+        // STACK: []
+
+        // Create and store an index into a local.
+        let index = generator.module.locals.add(ValType::I32);
+        builder.i32_const(0);
```

Comment:
You can just use an i64 here, so you don't need to extend it later. I don't think we're concerned about running fast on 32-bit devices.

Reply:
Done.